src/Uno.UI.Runtime.Skia.X11/X11NativeElementHostingExtension.cs

```diff
@@ -0,0 +1,285 @@
+	private Rect? _lastFinalRect;
+	private Rect? _lastArrangeRect;
+	private Rect? _lastClipRect;
+	private bool _layoutDirty = true;
+	private bool? _xShapesPresent;
+	private readonly ContentPresenter _presenter;
+	private readonly IntPtr _display;
+
+	public X11NativeElementHostingExtension(ContentPresenter contentPresenter)
+	{
+		_presenter = contentPresenter;
+		_display = ((X11XamlRootHost)X11Manager.XamlRootMap.GetHostForRoot(_presenter.XamlRoot!)!).RootX11Window.Display;
+	}
+
+	private XamlRoot? XamlRoot => _presenter.XamlRoot;
+
+	internal static IEnumerable<XRectangle> GetNativeElementRects(X11XamlRootHost host)
+	{
+		if (_hostToNativeElementHosts.TryGetValue(host, out var set))
+		{
+			foreach (var hostingExtension in set)
+			{
+				if (hostingExtension._lastFinalRect is { } rect)
+				{
+					yield return new XRectangle
+					{
+						X = (short)rect.X,
+						Y = (short)rect.Y,
+						H = (short)rect.Height,
+						W = (short)rect.Width
+					};
+				}
+			}
+		}
+	}
+
+	public bool IsNativeElement(object content)
+	{
+		if (content is not X11NativeWindow nativeWindow)
+		{
+			return false;
+		}
+
+		using var _1 = X11Helper.XLock(_display);
+
+		var _3 = XLib.XQueryTree(_display, XLib.XDefaultRootWindow(_display), out IntPtr root, out _, out var children, out _);
+		var _4 = XLib.XFree(children);
+
+		// _NET_CLIENT_LIST only identifies top-level windows, not subwindows.
+		var status = XLib.XGetWindowProperty(
+			_display,
+			root,
+			X11Helper.GetAtom(_display, X11Helper._NET_CLIENT_LIST),
+			0,
+			new IntPtr(0x7fffffff),
+			false,
+			X11Helper.AnyPropertyType,
+			out _,
+			out _,
+			out var length,
+			out _,
+			out IntPtr windowArray);
+
+		if (status == X11Helper.Success)
+		{
+			unsafe
+			{
+				var span = new Span<IntPtr>(windowArray.ToPointer(), (int)length);
+				foreach (var window in span)
+				{
+					if (window == nativeWindow.WindowId)
+					{
+						return true;
+					}
+				}
+			}
+		}
+
+		return FindWindowById(_display, nativeWindow.WindowId, root) != IntPtr.Zero;
+	}
+	public void AttachNativeElement(object content)
+	{
+		if (content is X11NativeWindow nativeWindow
+			&& XamlRoot is { } xamlRoot
+			&& X11Manager.XamlRootMap.GetHostForRoot(xamlRoot) is X11XamlRootHost host)
+		{
+			using var _1 = X11Helper.XLock(_display);
+
+			// this seems to be necessary or else the WM will keep detaching the subwindow
+			XWindowAttributes attributes = default;
+			var _2 = XLib.XGetWindowAttributes(_display, nativeWindow.WindowId, ref attributes);
+			attributes.override_direct = /* True */ 1;
+
+			unsafe
+			{
+				IntPtr attr = Marshal.AllocHGlobal(Marshal.SizeOf(attributes));
+				Marshal.StructureToPtr(attributes, attr, false);
+				var _3 = X11Helper.XChangeWindowAttributes(_display, nativeWindow.WindowId, (IntPtr)XCreateWindowFlags.CWOverrideRedirect, (XSetWindowAttributes*)attr.ToPointer());
+				Marshal.FreeHGlobal(attr);
```

Comment:
This should be in a finally

Reply:
It shouldn't matter since `XChangeWindowAttributes` doesn't throw exceptions. If the call is incorrect for some reason, we will just crash.